# Семинар 3. Цикл while

## Как использовать условия с максимальной эффективностью?

### Применяйте правильно логику

Распространенной ошибкой при проверке нескольких условий является неправильное применение логики.

Этот пример пытается проверить , если две переменные имеют больше 2. Заявление оценивается как - if (a) and (b > 2).Это дает неожиданный результат , так как bool(a) вычисляется как True , когда не равен нулю. a

```
>>> a = 1
>>> b = 6
>>> if a and b > 2:
...     print('yes')
... else:
...     print('no')

yes
```

Каждую переменную нужно сравнивать отдельно.

```
>>> if a > 2 and b > 2:
...     print('yes')
... else:
...     print('no')

no
```

Другая, аналогичная ошибка допущена при проверке, является ли переменная одним из нескольких значений. Утверждение в этом примере оценивается как - if (a == 3) or (4) or (6).Это дает неожиданный результат , так как bool(4) и bool(6) каждый вычисляться значение True

```
>>> a = 1
>>> if a == 3 or 4 or 6:
...     print('yes')
... else:
...     print('no')

yes
```

Опять каждое сравнение должно быть сделано отдельно

```
>>> if a == 3 or a == 4 or a == 6:
...     print('yes')
... else:
...     print('no')

no
```

# Синтаксис цикла while и его применение

Пусть у нас есть небольшое производство восковых свечей. Однако мы понимаем, что производить свечи можно только до тех пор, пока не кончится воск. Для таких случаев и существует **цикл while**.

    Слово while в переводе с английского означает «пока», поэтому, как правило, этот вид цикла используют в задачах с граничными условиями: повторять цикл необходимо, пока что-то не кончилось, пока что-то не произошло, и тому подобное.
    
Чтобы рассчитать количество свечей, которые можно отлить из купленного воска, необходимо разобраться, как работать с циклом **while**.


#### Рассмотрим конструкцию цикла while.

СИНТАКСИС

In [ ]:
while <логическое выражение>:
    <тело цикла>

Разберем синтаксис подробно:

|ЭЛЕМЕНТ|ОПИСАНИЕ|
|:---:|:---:|
|<font size="3">**while**</font>|<font size="3">Зарезервированное ключевое слово, с которого начинается цикл|
|<font size="3">**<логическое выражение>**	| <font size="3">Выражение, результатом которого является булево значение `True` или `False`. <br>Если результат <логического выражения> `True`, то цикл продолжает работу. <br>Как только результат <логического выражения> поменялся на `False`, цикл останавливается|
|<font size="3">**:**|<font size="3">Зарезервированный ключевой символ. Обязательно после указания <логического выражения> использовать двоеточие|
|<font size="3">**<тело цикла>**|<font size="3">Инструкции, которые выполняются при каждой итерации. Записывается с отступом относительно ключевого слова while|

А теперь попробуем решить нашу задачу. Для этого мы на каждом этапе цикла будем увеличивать количество изготовленных свечей на 1, а массу воска уменьшать на вес одной свечи, до тех пор пока масса оставшегося воска не станет меньше массы свечи, то есть оставшегося воска не хватит для производства одной свечи.

In [ ]:
wax_gr = 5000 #начальное кол-во грамм купленного воска

candle_wax_gr = 120 #грамм воска необходимо для изготовления свечи

candle_cnt = 0 #введем переменную количества свечей (первоначально равна 0)

while wax_gr > candle_wax_gr:

    candle_cnt += 1

    wax_gr -= candle_wax_gr

    print('Уже сделано: ', candle_cnt,' свечей,', 'осталось ',wax_gr, ' грамм воска')

Уже сделано:  1  свечей, осталось  4880  грамм воска
Уже сделано:  2  свечей, осталось  4760  грамм воска
Уже сделано:  3  свечей, осталось  4640  грамм воска
Уже сделано:  4  свечей, осталось  4520  грамм воска
Уже сделано:  5  свечей, осталось  4400  грамм воска
Уже сделано:  6  свечей, осталось  4280  грамм воска
Уже сделано:  7  свечей, осталось  4160  грамм воска
Уже сделано:  8  свечей, осталось  4040  грамм воска
Уже сделано:  9  свечей, осталось  3920  грамм воска
Уже сделано:  10  свечей, осталось  3800  грамм воска
Уже сделано:  11  свечей, осталось  3680  грамм воска
Уже сделано:  12  свечей, осталось  3560  грамм воска
Уже сделано:  13  свечей, осталось  3440  грамм воска
Уже сделано:  14  свечей, осталось  3320  грамм воска
Уже сделано:  15  свечей, осталось  3200  грамм воска
Уже сделано:  16  свечей, осталось  3080  грамм воска
Уже сделано:  17  свечей, осталось  2960  грамм воска
Уже сделано:  18  свечей, осталось  2840  грамм воска
Уже сделано:  19  свечей, осталось  2

Рассмотрим работу написанного нами цикла на каждой итерации:

|Номер итерации	|Кол-во свечей	|Кол-во оставшегося воска|Кол-во оставшегося воска>120 грамм?|Цикл продолжается?|
|:---:|:---:|:---:|:---:|:---:|
|1|	1|	4880|	True|	Да|
|2|	2|	4760|	True|	Да|
|3|	3|	4640|	True|	Да|
|4|	4|	4520|	True|	Да|
|5|	5|	4400|	True|	Да|
|. . . . . . . .|	. . . . . . . . . . . .| 	. . . . . . . . . . . . . . . . .|	. . . . . . . . . . . . . . . . .|	. . . . . . . . . . . . .|
|37|	37|	560|	True|	Да|
|38|	38|	440|	True|	Да|
|39|	39|	320|	True|	Да|
|40|	40|	200|	True|	Да|
|41|	41|	80|	False|	Нет|

Как только цикл дошел до массы воска в 80 грамм, он остановился, поскольку дальше логическое утверждение принимает значение False. Казалось бы, зачем мы все это выдумывали, если можно было 5 кг разделить на 120 г и получить 41.67, то есть 41 целую свечку и еще немножко воска.

А что, если процесс гораздо сложнее? Например, изготавливаются свечи разного диаметра, требующие разное количество воска, и их необходимо чередовать при производстве, легко ли тогда ответить на вопрос, сколько свечей каждого вида можно произвести? Поэтому для работы автоматизированных задач зачастую используют цикл while или его аналоги в других языках программирования.

    При подсчете количества изготовленных свечей мы на каждой итерации прибавляли к переменной candle_cnt единицу. Такой подход часто используют в работе с любыми циклами, а переменная, которую формируют таким образом (в нашем случае candle_cnt), называется счетчиком.

## <center> Время решать задачи!</center>
<img src="https://media0.giphy.com/media/5Az0uB1iLeuvgcOoFK/giphy.gif?cid=ecf05e47mwin3v49qiaktg6pk3qmmss17kdie27astq2512a&ep=v1_gifs_search&rid=giphy.gif&ct=g" width=300>

### Задача 1

Пользователь вводит число. Ваша задача распечатать на каждой строке звездочки от одной до введенного пользователем числа. Например, если пользователь ввел 5, то ваша программа должна вывести:

    *
    **
    ***
    ****
    *****

In [ ]:
# ваш код

### Задача 2.1

Напишите программу, которая будет проверять верность введенного пользователем пароля, пока пароль не совпадет с истинным.

Верный пароль пользователя – PythonLanguage345

In [ ]:
correct_passw = 'PythonLanguage345'

# ваш код

### Задача 2.2

Усложним задачу. У пользователя теперь есть только 3 попытки на ввод пароля, и программа должна сообщать ему об этом после каждой неудачной попытки.

In [ ]:
correct_passw = 'PythonLanguage345'
max_attempt = 3
print(f'У вас есть {max_attempt} попытки, чтобы ввести пароль')

# ваш код


У вас есть 3 попытки, чтобы ввести пароль


### Задача 3

Напишите программу через цикл `while True`, которая будет запрашивать у пользователя, что ему нужно купить сегодня в магазине, пока пользователь не введет "стоп".

После этого программа должна вывести введенные продукты через пробел.

яблоко морковь молоко стоп

In [ ]:
# ваш код

### Задача 4

Список Мюнхгаузена

Помогите Барону Мюнхгаузену составить список дел на день. Известно, что каждое дело записывается в виде отдельной фразы, например 'Полет на Луну', а последним делом должна быть обязательно 'Война с Англией'. В результате должен получиться список дел через запятую и без пробелов между фразами.

**Для тестирования используйте следующие данные:**

Подъем \
Разгон облаков \
Полет на Луну \
Разговор с Ньютоном \
Арест \
Война с Англией

**Результатом должна стать строка:**

Подъем,Разгон облаков,Полет на Луну,Разговор с Ньютоном,Арест,Война с Англией

In [ ]:
tasks = '' #строка, в которой будут все дела через запятую и пробел

# ваш код

### Задача 5

Продолжаем писать письма всем участникам нашего курса.

В прошлый раз мы написали код, который автоматически подставляет верное обращение "Дорогой Имя Отчество!" или "Дорогая Имя Отчество!" (суть была в проверке последней буквы отчества. Отчество мужчины заканчивается буквой "ч").

Но наша программа отрабатывала только один раз. Мы хотим дописать ее таким образом, чтобы она постоянно у нас спрашивала ИО на вход, печатала письмо, снова спрашивала ИО и печатала письмо, пока мы не напишем "КОНЕЦ".

In [ ]:
letter = '''Прости мне, милый друг,
Двухлетнее молчанье:
Писать тебе посланье
Мне было недосуг.
На тройке пренесенный
Из родины смиренной
В великий град Петра,
От утра до утра
Два года всё кружился
Без дела в хлопотах,
Зевая, веселился.
'''

In [ ]:
# нужно обязательно запустить ячейку выше, чтобы в память записалась переменная letter
# доделаем код с помощью while True

### Задача 5

Пользователь вводит слово, вам нужно напечатать каждую букву этого слова с новой строки. Для решения используйте цикл while.

Пример:

Волга

В \
о \
л \
г \
а

In [ ]:
# ваш код

### Задача 6 (формат B2 НЭ)

Реализуйте программу, которая проверяет, сколько слов начинаются с заглавной буквы:

* Задана переменная counter, которая равна нулю. Также в строке letters записаны все заглавные буквы английского алфавита. Этот код уже написан.
* В цикле while считываются строки до тех пор, пока не введена строка "END" (не учитывается в последовательности).
* Для каждого вводимого числа сделайте проверку: если строка начинается с заглавной буквы — прибавьте к счетчику 1.
* Гарантируется, что до слова "END" будет введено как минимум одна строка.
* Напечатайте число, записанное в переменную counter.

In [ ]:
counter = 0
letters = "ABCDEFGHIJKLMNOPQRSTUVXYZ"

# ваш код ниже

# Применение цикла while в практических задачах

На практике цикл while, как правило, решает круг довольно узких задач. Давайте рассмотрим их подробнее.

1. Загрузка данных с ограничениями

Многие из вас наверняка работали с Microsoft Excel. Эта программа имеет ограничение на количество строк, хранящихся на одном листе, — примерно 1 млн записей. Такое ограничение не заметно, если вы работаете с небольшими объемами информации, однако становится критичным, если массивы данных измеряются гигабайтами. Поэтому при заливке данных в Excel разработчики часто используют цикл while: пока количество записей на листе меньше 1 млн, данные загружаются, а в случае превышения лимита загрузка данных будет продолжаться на следующем листе.

Код
```
rows = 0

while rows < 1000000:

   save_data  # сохраняем данные на лист

   rows = rows + count(data)  # обновляем количество уже сохраненных записей
```

Загрузка данных в Microsoft Excel — это один из множества примеров задачи загрузки данных куда-либо. Ограничение может возникнуть в количестве записей, файлов, секунд,  объеме информации и так далее. Подавляющее большинство этих задач решаются с помощью цикла while.


2. Технологические процессы с ограничениями

При разборе задач на синтаксис while мы рассматривали процесс изготовления свечей, и для упрощения базового синтаксиса задача сводилась к одному ограничению — количеству воска, однако на практике таких переменных в разы больше, а на больших производствах их количество и вовсе может достигать нескольких сотен. Очевидно, что в таких условиях польза от вычисления искомых значений, а также контроль изготовления продукции на предприятиях с помощью компьютерных программ неоценимы, поэтому программисты часто используют этот тип цикла в практических задачах, связанных с производством.

3. Программы мониторинга

Программы мониторинга служат примером программ с использованием бесконечных циклов. В настоящее время существует множество систем, которые работают без контроля человека. Некоторые из них выполняют свои задачи строго по расписанию — их настроили на работу в определенное время и они ее в это время выполнили. Однако существуют и другие программы — они в случайный момент времени «оживают», проверяют, что вокруг них происходит, выполняют свою работу и снова «засыпают». Такие системы называются системами мониторинга.

Ваша Алиса не работает, пока вы не назвали ее по имени. Все остальное время она "спит".

Как же такие системы узнают, что для них появилась работа? На самом деле никак, это и есть пример бесконечного цикла. Система работает все время: если для нее есть работа, она ее выполняет, если нет, то система проверяет, не появилась ли работа, и так бесконечно.

```
a = 5

while a > 4:

    if find_new_job: #если найдено новое задание

        make_new_job #делай это задание

    else:

        pass
```
        
find_new_job и make_new_job — это «говорящие» имена функций, ранее мы их нигде не указывали.

Как правило, нет необходимости проверять наличие новых заданий каждую секунду. В этом случае для экономии ресурсов программу искусственно «замораживают» с помощью специальных функций заморозки. В Python зачастую используют функцию sleep библиотеки time.